nmostesting/mocks/Registry.py

```diff
@@ -271,3 +417,192 @@ def heartbeat(version, node_id):
+    registry = REGISTRIES[flask.current_app.config["REGISTRY_INSTANCE"]]
+    if not registry.enabled:
+        abort(503)
+    authorized = registry.check_authorized(request.headers, request.path)
+    if authorized is not True:
+        abort(authorized)
+
+    MIN_SINCE = "0:0"
+    MAX_UNTIL = "9999999999:9999999"
+
+    base_data = []
+
+    if request.args.get('paging.since') or request.args.get('paging.until'):
+        registry.pagination_used = True
+
+    since = request.args.get('paging.since') if request.args.get('paging.since') else MIN_SINCE
+    until = request.args.get('paging.until') if request.args.get('paging.until') else MAX_UNTIL
+    new_until = until
+
+    try:
+        valid_resource_types = ['device', 'flow', 'node', 'receiver', 'sender', 'source', 'subscription']
+
+        resource_type = resource.rstrip("s")
+
+        if resource_type not in valid_resource_types:
+            error_message = {"code": 404,
+                             "error": "Invalid resource",
+                             "debug": resource_type}
+            return Response(json.dumps(error_message), status=404, mimetype='application/json')
+
+        registry.query_api_called = True
+
+        # Check to see if resource is being requested as a query
+        if request.args.get('id'):
+            resource_id = request.args.get('id')
+            base_data.append(registry.get_resources()[resource_type][resource_id])
+        else:
+            data = registry.get_resources()[resource_type]
+
+            # only paginate for version v1.3 and up
+            if NMOSUtils.compare_api_version("v1.1", version) > 0:
+                for key, value in data.items():
+                    base_data.append(value)
+            else:
+                data_list = [d for k, d in data.items()]
+                sorted_list = sorted(data_list, key=functools.cmp_to_key(compare_resources))
+
+                result_count = 0
+                for value in sorted_list:
+                    if NMOSUtils.compare_resource_version(value['version'], since) >= 0 \
+                            and NMOSUtils.compare_resource_version(until, value['version']) > 0:
+                        if result_count < registry.paging_limit:
+                            base_data.append(value)
+                            result_count += 1
+                        else:
+                            new_until = value['version']
+                            break
+
+    except Exception:
+        pass
+
+    response = Response(json.dumps(base_data), mimetype='application/json')
+
+    # add pagination headers for v1.1 and up
+    if NMOSUtils.compare_api_version("v1.1", version) <= 0:
+        link = ""
+        if new_until != MAX_UNTIL:
+            link += "<http://localhost:5102/x-nmos/query/v1.3/" + resource_type + "s/?paging.since=" + new_until \
+                + "&paging.limit=" + str(registry.paging_limit) + ">; rel=\"next\""
+
+        if since != MIN_SINCE:
```

Comment:
No need to exclude `first`, it can always be present, even if it's the same as `prev`.

Reply:
Ah, you probably meant this to be `prev` actually?

In which case, fix that, and whether to add `first` and `last` is a separate question.

Reply:
Good catch. 'first' changed to 'prev'